### PR TITLE
[SLA] PIM-5194: Performance issues on families

### PR DESCRIPTION
**Pull request by solivier**
| Q                 | A
| ----------------- | ---
| Specs             | Yes
| Behats            | No
| Blue CI           | yes
| Changelog updated | Yes
| Review and 2 GTM  | 2 GTM

In order to be able to load the PEF with more than 6k families, we need to ajaxify the change family select2. Thanks @sumbobyboys for the help ;)

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php [hunk 1]
```diff
@@ -12,6 +15,8 @@
  * @author    Julien Sanchez <julien@akeneo.com>
  * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
  * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ *
+ * TODO: to rework on master
```
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php]: remove this comment

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php [hunk 2]
```diff
@@ -36,9 +49,17 @@ public function __construct(EntityRepository $familyRepository, NormalizerInterf
      *
      * @return JsonResponse
      */
-    public function indexAction()
+    public function indexAction(Request $request)
     {
-        $families = $this->familyRepository->findAll();
+        #TODO: to rework on master, drop the if condition
+        if ($this->familySearchableRepo !== null) {
```
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php]: yoda style

File: src/Pim/Bundle/EnrichBundle/Entity/Repository/FamilySearchableRepository.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        $this->entityName    = $entityName;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findBySearch($search = null, array $options = [])
+    {
+        $qb = $this->entityManager->createQueryBuilder()->select('f')->from($this->entityName, 'f');
+        $qb->select("f");
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Entity/Repository/FamilySearchableRepository.php]: single quote please

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/base-fetcher.js [hunk 4]
```diff
@@ -62,7 +62,7 @@ define(['jquery', 'underscore', 'backbone', 'routing'], function ($, _, Backbone
                     }).fail(function () {
                         console.log(arguments);
 
-                        return deferred.reject();
+                        return deferred.resolve(null);
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/base-fetcher.js]: we can drop that now :)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/manager/attribute-manager.js [hunk 5]
```diff
@@ -84,7 +94,7 @@ define([
             },
 
             /**
-             * Generate a single value for the given attribute, scope and lcoale
+             * Generate a single value for the given attribute, scope and locale
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/manager/attribute-manager.js]: :+1:

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/attributes.js [hunk 6]
```diff
@@ -86,9 +86,8 @@ define(
                 this.getConfig().then(function () {
                     var product = this.getFormData();
                     $.when(
-                        FetcherRegistry.getFetcher('family').fetchAll(),
                         ProductManager.getValues(product)
-                    ).then(function (families, values) {
+                    ).then(function (values) {
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/attributes.js]: much cleaner !

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/attributes.js [hunk 7]
```diff
@@ -132,24 +131,26 @@ define(
...
-                        scope: scope.code,
-                        scopeLabel: scope.label,
-                        uiLocale: UserContext.get('catalogLocale'),
-                        optional: AttributeManager.isOptional(field.attribute, product, families),
-                        removable: SecurityContext.isGranted('pim_enrich_product_remove_attribute')
-                    });
-                    field.setValues(values);
+            renderField: function (product, attributeCode, values) {
+                return FieldManager.getField(attributeCode).then(function (field) {
+                    return $.when(
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/attributes.js]: we can rework that IRL :)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/meta/change-family.js [hunk 8]
```diff
@@ -33,35 +34,65 @@ define(
...
+                        familyModal.close();
                     }.bind(this));
-
-                    familyModal.open();
-                    familyModal.$('select').select2({ allowClear: true });
-                    familyModal.$('.modal-body').css({'line-height': '25px', 'height': 130});
                 }.bind(this));
+
+                familyModal.open();
+                familyModal.$('.family-select2').select2('destroy').select2({
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/meta/change-family.js]: not sure that we need to destroy it.

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/meta/change-family.js [hunk 9]
```diff
@@ -33,35 +34,65 @@ define(
...
+                    ajax: {
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/meta/change-family.js]: does the default value works ? Curious about how you did it :)

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php [hunk 10]
```diff
@@ -47,4 +68,25 @@ public function indexAction()
...
+     *
+     * @param int $identifier
+     *
+     * @return JsonResponse
+     */
+    public function getAction($identifier)
+    {
+        $family = $this->familyRepository->findOneByCode($identifier);
+
+        if (!$family) {
```
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php]: null !==

**Comment by sumbobyboys** [comment]
Really nice javascript PR ! Much cleaner and fast !

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php (hunk 10 above)
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php]: !$var is for boolean!

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php [hunk 11]
```diff
@@ -47,4 +68,25 @@ public function indexAction()
...
+    /**
+     * Get a single family
+     *
+     * @param int $identifier
+     *
+     * @return JsonResponse
+     */
+    public function getAction($identifier)
+    {
+        $family = $this->familyRepository->findOneByCode($identifier);
```
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php]: findOneByIdentifier ?

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php [hunk 12]
```diff
@@ -47,4 +68,25 @@ public function indexAction()
...
+
+        if (!$family) {
+            throw new NotFoundHttpException(sprintf('Family with code "%s" not found', $identifier));
+        }
+
+        $normalizedFamily = [];
+        return new JsonResponse(
```
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php]: add blank line before return please

File: src/Pim/Bundle/EnrichBundle/Entity/Repository/FamilySearchableRepository.php (hunk 3 above)
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Entity/Repository/FamilySearchableRepository.php]: remove it or remove your previous `select('f')`

File: src/Pim/Bundle/EnrichBundle/Entity/Repository/FamilySearchableRepository.php [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+        $qb->join('f.translations', 'ft');
+
+        if (null !== $search) {
+            $qb->where("f.code like :search")->setParameter('search', "%$search%");
```
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Entity/Repository/FamilySearchableRepository.php]: simple quote please

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php (hunk 1 above)
Comment (solivier) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php]: why ?

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php (hunk 11 above)
Comment (solivier) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php]: family repository does not have this method

File: src/Pim/Bundle/EnrichBundle/Entity/Repository/FamilySearchableRepository.php [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+use Doctrine\ORM\EntityManager;
+
+/**
+ * Family searchable repository
+ *
+ * @author    Olivier Soulet <olivier.soulet@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class FamilySearchableRepository implements SearchableRepositoryInterface
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Entity/Repository/FamilySearchableRepository.php]: should be in `Pim\Bundle\EnrichBundle\Doctrine\ORM\Repository`

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php (hunk 1 above)
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php]: weird to see that..

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php (hunk 11 above)
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php]: https://github.com/akeneo/pim-community-dev/blob/1.4/src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/FamilyRepository.php#L207

:)

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php [hunk 15]
```diff
@@ -47,4 +68,26 @@ public function indexAction()
...
+    /**
+     * Get a single family
+     *
+     * @param int $identifier
+     *
+     * @return JsonResponse
+     */
+    public function getAction($identifier)
+    {
+        $family = $this->familyRepository->findOneBy(['code' => $identifier]);
```
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php]: findOneByIdentifier ?

Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php]: You can use `findOneByIdentifier`
https://github.com/akeneo/pim-community-dev/blob/master/src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/FamilyRepository.php#L207

File: src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilySearchableRepository.php [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+     * {@inheritdoc}
+     */
+    public function findBySearch($search = null, array $options = [])
+    {
+        $qb = $this->entityManager->createQueryBuilder()->select('f')->from($this->entityName, 'f');
+        $qb->join('f.translations', 'ft');
+
+        if (null !== $search) {
+            $qb->where('f.code like :search')->setParameter('search', "%$search%");
+            $qb->orWhere("ft.label like :search")->setParameter('search', "%$search%");
```
Comment (skeleton) [inline src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilySearchableRepository.php]: replace `"` by `'`

File: src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilySearchableRepository.php [hunk 17]
```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace Pim\Bundle\EnrichBundle\Doctrine\ORM\Repository;
+
+use Akeneo\Component\StorageUtils\Repository\SearchableRepositoryInterface;
+use Doctrine\ORM\EntityManager;
```
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilySearchableRepository.php]: Maybe a Doctrine/ORM/EntityManagerInterface ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/base-fetcher.js [hunk 18]
```diff
@@ -62,7 +62,7 @@ define(['jquery', 'underscore', 'backbone', 'routing'], function ($, _, Backbone
                     }).fail(function () {
                         console.log(arguments);
```
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/base-fetcher.js]: Is it a good practice ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/base-fetcher.js [hunk 19]
```diff
@@ -62,7 +62,7 @@ define(['jquery', 'underscore', 'backbone', 'routing'], function ($, _, Backbone
...
 
-                        return deferred.reject();
+                            return deferred.reject();
```
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/fetcher/base-fetcher.js]: Useless :)

File: src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilySearchableRepository.php (hunk 17 above)
Comment (jmleroux) [inline src/Pim/Bundle/EnrichBundle/Doctrine/ORM/Repository/FamilySearchableRepository.php]: :+1: for @pierallard

**Comment by jjanvier** [comment]
GTM (but I didn't read JS)

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php [hunk 20]
```diff
@@ -47,4 +70,26 @@ public function indexAction()
...
+        $family = $this->familyRepository->findOneByIdentifier($identifier);
+
+        if (null === $family) {
+            throw new NotFoundHttpException(sprintf('Family with code "%s" not found', $identifier));
+        }
+
+        $normalizedFamily = [];
+
+        return new JsonResponse(
+            $normalizedFamily[$family->getCode()] = $this->normalizer->normalize($family, 'json')
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/FamilyController.php]: nope we should not do that. It's not normal

File: src/Pim/Bundle/EnrichBundle/Resources/public/spec/manager/AttributeManagerSpec.js [hunk 21]
```diff
@@ -0,0 +1,80 @@
+/* global describe, it, expect, spyOn, beforeEach */
+'use strict';
+
+define(
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/spec/manager/AttributeManagerSpec.js]: this file is not really simple, could you explain it ? :trollface:

**Comment by juliensnz** [comment]
GTM

**Pull request merged**